File: backend/src/main/java/com/hrms/application/contract/service/ContractService.java
package com.hrms.application.contract.service;

import com.hrms.api.contract.dto.*;
import com.hrms.common.exception.ResourceNotFoundException;
import com.hrms.common.security.SecurityContext;
import com.hrms.common.security.TenantContext;
import com.hrms.domain.contract.*;
import com.hrms.domain.employee.Employee;
import com.hrms.infrastructure.contract.repository.*;
import com.hrms.application.employee.service.EmployeeService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Service for managing contracts
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class ContractService {

    private final ContractRepository contractRepository;
    private final ContractVersionRepository versionRepository;
    private final ContractSignatureRepository signatureRepository;
    private final ContractReminderRepository reminderRepository;
    private final EmployeeService employeeService;

    // ===================== CRUD Operations =====================

    /**
     * Create a new contract
     */
    @Transactional
    public ContractDto createContract(CreateContractRequest request) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        UUID userId = SecurityContext.getCurrentUserId();

        Contract contract = Contract.builder()
                .tenantId(tenantId)
                .title(request.getTitle())
                .type(request.getType())
                .status(ContractStatus.DRAFT)
                .employeeId(request.getEmployeeId())
                .vendorName(request.getVendorName())
                .startDate(request.getStartDate())
                .endDate(request.getEndDate())
                .autoRenew(request.getAutoRenew() != null ? request.getAutoRenew() : false)
                .renewalPeriodDays(request.getRenewalPeriodDays())
                .value(request.getValue())
                .currency(request.getCurrency() != null ? request.getCurrency() : "USD")
                .description(request.getDescription())
                .terms(request.getTerms())
                .documentUrl(request.getDocumentUrl())
                .createdBy(userId)
                .build();

        contract = contractRepository.save(contract);
        log.info("Contract created: {} ({})", contract.getId(), contract.getTitle());

        // Create initial version
        createVersion(contract.getId(), 1, request.getTerms(), "Initial version");

        return toDto(contract);
    }

    /**
     * Update an existing contract
     */
    @Transactional
    public ContractDto updateContract(UUID contractId, UpdateContractRequest request) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        UUID userId = SecurityContext.getCurrentUserId();

        Contract contract = contractRepository.findByIdAndTenantId(contractId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Contract not found"));

        // Update fields
        if (request.getTitle() != null) contract.setTitle(request.getTitle());
        if (request.getType() != null) contract.setType(request.getType());
        if (request.getStatus() != null) contract.setStatus(request.getStatus());
        if (request.getEmployeeId() != null) contract.setEmployeeId(request.getEmployeeId());
        if (request.getVendorName() != null) contract.setVendorName(request.getVendorName());
        if (request.getStartDate() != null) contract.setStartDate(request.getStartDate());
        if (request.getEndDate() != null) contract.setEndDate(request.getEndDate());
        if (request.getAutoRenew() != null) contract.setAutoRenew(request.getAutoRenew());
        if (request.getRenewalPeriodDays() != null) contract.setRenewalPeriodDays(request.getRenewalPeriodDays());
        if (request.getValue() != null) contract.setValue(request.getValue());
        if (request.getCurrency() != null) contract.setCurrency(request.getCurrency());
        if (request.getDescription() != null) contract.setDescription(request.getDescription());
        if (request.getTerms() != null) contract.setTerms(request.getTerms());
        if (request.getDocumentUrl() != null) contract.setDocumentUrl(request.getDocumentUrl());

        contract.setLastModifiedBy(userId);
        contract = contractRepository.save(contract);

        // Create new version if terms changed
        if (request.getTerms() != null) {
            Integer nextVersion = getNextVersionNumber(contractId);
            createVersion(contractId, nextVersion, request.getTerms(), "Updated by user");
        }

        log.info("Contract updated: {} ({})", contract.getId(), contract.getTitle());
        return toDto(contract);
    }

    /**
     * Get contract by ID
     */
    @Transactional(readOnly = true)
    public ContractDto getContractById(UUID contractId) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        Contract contract = contractRepository.findByIdAndTenantId(contractId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Contract not found"));
        return toDtoWithSignatures(contract);
    }

    /**
     * Get all contracts for tenant
     */
    @Transactional(readOnly = true)
    public Page<ContractListDto> getAllContracts(Pageable pageable) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findByTenantId(tenantId, pageable)
                .map(this::toListDto);
    }

    /**
     * Get contracts by status
     */
    @Transactional(readOnly = true)
    public Page<ContractListDto> getContractsByStatus(ContractStatus status, Pageable pageable) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findByTenantIdAndStatus(tenantId, status, pageable)
                .map(this::toListDto);
    }

    /**
     * Get contracts by type
     */
    @Transactional(readOnly = true)
    public Page<ContractListDto> getContractsByType(ContractType type, Pageable pageable) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findByTenantIdAndType(tenantId, type, pageable)
                .map(this::toListDto);
    }

    /**
     * Get contracts for an employee
     */
    @Transactional(readOnly = true)
    public Page<ContractListDto> getEmployeeContracts(UUID employeeId, Pageable pageable) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findByTenantIdAndEmployeeId(tenantId, employeeId, pageable)
                .map(this::toListDto);
    }

    /**
     * Search contracts
     */
    @Transactional(readOnly = true)
    public Page<ContractListDto> searchContracts(String search, Pageable pageable) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.searchContracts(tenantId, search, pageable)
                .map(this::toListDto);
    }

    /**
     * Delete contract
     */
    @Transactional
    public void deleteContract(UUID contractId) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        Contract contract = contractRepository.findByIdAndTenantId(contractId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Contract not found"));
        contractRepository.delete(contract);
        log.info("Contract deleted: {}", contractId);
    }

    // ===================== Status Transitions =====================

    /**
     * Mark contract as pending review
     */
    @Transactional
    public ContractDto markAsPendingReview(UUID contractId) {
        Contract contract = getContractEntity(contractId);
        contract.setStatus(ContractStatus.PENDING_REVIEW);
        contractRepository.save(contract);
        return toDto(contract);
    }

    /**
     * Mark contract as pending signatures
     */
    @Transactional
    public ContractDto markAsPendingSignatures(UUID contractId) {
        Contract contract = getContractEntity(contractId);
        contract.setStatus(ContractStatus.PENDING_SIGNATURES);
        contractRepository.save(contract);
        return toDto(contract);
    }

    /**
     * Mark contract as active
     */
    @Transactional
    public ContractDto markAsActive(UUID contractId) {
        Contract contract = getContractEntity(contractId);
        contract.markAsActive();
        contractRepository.save(contract);
        return toDto(contract);
    }

    /**
     * Mark contract as terminated
     */
    public ContractDto terminateContract(UUID contractId) {
        Contract contract = getContractEntity(contractId);
        contract.markAsTerminated();
        contractRepository.save(contract);
        log.info("Contract terminated: {}", contractId);
        return toDto(contract);
    }

    /**
     * Renew contract
     */
    public ContractDto renewContract(UUID contractId) {
        Contract contract = getContractEntity(contractId);
        if (contract.getEndDate() != null && contract.getRenewalPeriodDays() != null) {
            LocalDate newEndDate = contract.getEndDate().plusDays(contract.getRenewalPeriodDays());
            contract.setEndDate(newEndDate);
            contract.setStatus(ContractStatus.RENEWED);
            contractRepository.save(contract);
            log.info("Contract renewed: {} (new end date: {})", contractId, newEndDate);
        }
        return toDto(contract);
    }

    // ===================== Expiry and Status Checks =====================

    /**
     * Get expiring contracts
     */
    @Transactional(readOnly = true)
    public List<ContractListDto> getExpiringContracts(int days) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        LocalDate today = LocalDate.now();
        LocalDate expiryDate = today.plusDays(days);

        return contractRepository.findExpiringContracts(tenantId, ContractStatus.ACTIVE, today, expiryDate)
                .stream()
                .map(this::toListDto)
                .collect(Collectors.toList());
    }

    /**
     * Get expired contracts
     */
    @Transactional(readOnly = true)
    public List<ContractListDto> getExpiredContracts() {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findExpiredContracts(tenantId)
                .stream()
                .map(this::toListDto)
                .collect(Collectors.toList());
    }

    /**
     * Get active contracts
     */
    @Transactional(readOnly = true)
    public List<ContractListDto> getActiveContracts() {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findActiveContracts(tenantId)
                .stream()
                .map(this::toListDto)
                .collect(Collectors.toList());
    }

    // ===================== Version Management =====================

    /**
     * Get version history
     */
    @Transactional(readOnly = true)
    public List<Map<String, Object>> getVersionHistory(UUID contractId) {
        List<ContractVersion> versions = versionRepository.findByContractIdOrderByVersionNumberDesc(contractId);
        return versions.stream()
                .map(v -> Map.of(
                        "versionNumber", v.getVersionNumber(),
                        "changeNotes", v.getChangeNotes(),
                        "createdAt", v.getCreatedAt(),
                        "content", v.getContent()
                ))
                .collect(Collectors.toList());
    }

    // ===================== Helper Methods =====================

    private void createVersion(UUID contractId, Integer versionNumber, Map<String, Object> content, String changeNotes) {
        ContractVersion version = ContractVersion.builder()
                .contractId(contractId)
                .versionNumber(versionNumber)
                .content(content)
                .changeNotes(changeNotes)
                .createdBy(SecurityContext.getCurrentUserId())
                .build();
        versionRepository.save(version);
    }

    private Integer getNextVersionNumber(UUID contractId) {
        Integer maxVersion = versionRepository.findMaxVersionNumber(contractId);
        return (maxVersion != null ? maxVersion : 0) + 1;
    }

    private Contract getContractEntity(UUID contractId) {
        UUID tenantId = SecurityContext.getCurrentTenantId();
        return contractRepository.findByIdAndTenantId(contractId, tenantId)
                .orElseThrow(() -> new ResourceNotFoundException("Contract not found"));
    }

    private ContractDto toDto(Contract contract) {
        String employeeName = null;
        if (contract.getEmployeeId() != null) {
            try {
                Employee emp = employeeService.getByIdAndTenant(contract.getEmployeeId(), SecurityContext.getCurrentTenantId());
                employeeName = emp.getFirstName() + " " + emp.getLastName();
            } catch (Exception e) {
                log.debug("Could not load employee for contract: {}", contract.getEmployeeId());
            }
        }

        return ContractDto.builder()
                .id(contract.getId())
                .title(contract.getTitle())
                .type(contract.getType())
                .status(contract.getStatus())
                .employeeId(contract.getEmployeeId())
                .employeeName(employeeName)
                .vendorName(contract.getVendorName())
                .startDate(contract.getStartDate())
                .endDate(contract.getEndDate())
                .autoRenew(contract.getAutoRenew())
                .renewalPeriodDays(contract.getRenewalPeriodDays())
                .value(contract.getValue())
                .currency(contract.getCurrency())
                .description(contract.getDescription())
                .terms(contract.getTerms())
                .documentUrl(contract.getDocumentUrl())
                .createdBy(contract.getCreatedBy())
                .createdAt(contract.getCreatedAt())
                .updatedBy(contract.getLastModifiedBy())
                .updatedAt(contract.getUpdatedAt())
                .build();
    }

    private ContractDto toDtoWithSignatures(Contract contract) {
        ContractDto dto = toDto(contract);
        List<ContractSignature> signatures = signatureRepository.findByContractId(contract.getId());
        dto.setSignatureCount(signatures.size());
        dto.setPendingSignatureCount((int) signatures.stream()
                .filter(s -> s.getStatus() == SignatureStatus.PENDING)
                .count());
        dto.setSignatures(signatures.stream().map(this::toSignatureDto).collect(Collectors.toList()));
        return dto;
    }

    private ContractListDto toListDto(Contract contract) {
        String employeeName = null;
        if (contract.getEmployeeId() != null) {
            try {
                Employee emp = employeeService.getByIdAndTenant(contract.getEmployeeId(), SecurityContext.getCurrentTenantId());
                employeeName = emp.getFirstName() + " " + emp.getLastName();
            } catch (Exception e) {
                log.debug("Could not load employee for contract: {}", contract.getEmployeeId());
            }
        }

        int pendingSignatures = signatureRepository.findPendingSignatures(contract.getId()).size();

        return ContractListDto.builder()
                .id(contract.getId())
                .title(contract.getTitle())
                .type(contract.getType())
                .status(contract.getStatus())
                .employeeName(employeeName)
                .vendorName(contract.getVendorName())
                .startDate(contract.getStartDate())
                .endDate(contract.getEndDate())
                .value(contract.getValue())
                .currency(contract.getCurrency())
                .pendingSignatureCount(pendingSignatures)
                .createdAt(contract.getCreatedAt())
                .isExpiring(contract.isExpiringWithin(30))
                .isExpired(contract.isExpired())
                .build();
    }

    private ContractSignatureDto toSignatureDto(ContractSignature sig) {
        return ContractSignatureDto.builder()
                .id(sig.getId())
                .contractId(sig.getContractId())
                .signerId(sig.getSignerId())
                .signerName(sig.getSignerName())
                .signerEmail(sig.getSignerEmail())
                .signerRole(sig.getSignerRole())
                .status(sig.getStatus())
                .signedAt(sig.getSignedAt())
                .signatureImageUrl(sig.getSignatureImageUrl())
                .createdAt(sig.getCreatedAt())
                .build();
    }
}

File: frontend/lib/services/hrms/contract.service.ts
import { apiClient } from '../../api/client';
import {
  Contract,
  ContractListItem,
  CreateContractRequest,
  UpdateContractRequest,
  ContractSignature,
  SendForSigningRequest,
  ContractTemplate,
  CreateContractTemplateRequest,
  ContractStatus,
  ContractType,
  Page,
  SignatureSummary,
} from '../../types/hrms/contract';

class ContractService {
  // ===================== Contract CRUD =====================

  async createContract(data: CreateContractRequest): Promise<Contract> {
    const response = await apiClient.post<Contract>('/contracts', data);
    return response.data;
  }

  async getContractById(id: string): Promise<Contract> {
    const response = await apiClient.get<Contract>(`/contracts/${id}`);
    return response.data;
  }

  async getContracts(page: number = 0, size: number = 20): Promise<Page<ContractListItem>> {
    const response = await apiClient.get<Page<ContractListItem>>('/contracts', {
      params: { page, size },
    });
    return response.data;
  }

  async updateContract(id: string, data: UpdateContractRequest): Promise<Contract> {
    const response = await apiClient.put<Contract>(`/contracts/${id}`, data);
    return response.data;
  }

  async deleteContract(id: string): Promise<void> {
    await apiClient.delete(`/contracts/${id}`);
  }

  // ===================== Filtering & Search =====================

  async getContractsByStatus(status: ContractStatus, page: number = 0, size: number = 20): Promise<Page<ContractListItem>> {
    const response = await apiClient.get<Page<ContractListItem>>(`/contracts/status/${status}`, {
      params: { page, size },
    });
    return response.data;
  }

  async getContractsByType(type: ContractType, page: number = 0, size: number = 20): Promise<Page<ContractListItem>> {
    const response = await apiClient.get<Page<ContractListItem>>(`/contracts/type/${type}`, {
      params: { page, size },
    });
    return response.data;
  }

  async getEmployeeContracts(employeeId: string, page: number = 0, size: number = 20): Promise<Page<ContractListItem>> {
    const response = await apiClient.get<Page<ContractListItem>>(`/contracts/employee/${employeeId}`, {
      params: { page, size },
    });
    return response.data;
  }

  async searchContracts(search: string, page: number = 0, size: number = 20): Promise<Page<ContractListItem>> {
    const response = await apiClient.get<Page<ContractListItem>>('/contracts/search', {
      params: { search, page, size },
    });
    return response.data;
  }

  // ===================== Status Transitions =====================

  async markAsPendingReview(id: string): Promise<Contract> {
    const response = await apiClient.patch<Contract>(`/contracts/${id}/mark-pending-review`);
    return response.data;
  }

  async markAsPendingSignatures(id: string): Promise<Contract> {
    const response = await apiClient.patch<Contract>(`/contracts/${id}/mark-pending-signatures`);
    return response.data;
  }

  async markAsActive(id: string): Promise<Contract> {
    const response = await apiClient.patch<Contract>(`/contracts/${id}/mark-active`);
    return response.data;
  }

  async terminateContract(id: string): Promise<Contract> {
    const response = await apiClient.patch<Contract>(`/contracts/${id}/terminate`);
    return response.data;
  }

  async renewContract(id: string): Promise<Contract> {
    const response = await apiClient.patch<Contract>(`/contracts/${id}/renew`);
    return response.data;
  }

  // ===================== Expiry & Status Checks =====================

  async getExpiringContracts(days: number = 30): Promise<ContractListItem[]> {
    const response = await apiClient.get<ContractListItem[]>('/contracts/expiring', {
      params: { days },
    });
    return response.data;
  }

  async getExpiredContracts(): Promise<ContractListItem[]> {
    const response = await apiClient.get<ContractListItem[]>('/contracts/expired');
    return response.data;
  }

  async getActiveContracts(): Promise<ContractListItem[]> {
    const response = await apiClient.get<ContractListItem[]>('/contracts/active');
    return response.data;
  }

  // ===================== Version Management =====================

  async getVersionHistory(contractId: string): Promise<Record<string, unknown>[]> {
    const response = await apiClient.get<Record<string, unknown>[]>(`/contracts/${contractId}/versions`);
    return response.data;
  }

  // ===================== Signature Management =====================

  async sendForSigning(contractId: string, data: SendForSigningRequest): Promise<ContractSignature> {
    const response = await apiClient.post<ContractSignature>(`/contracts/${contractId}/send-for-signing`, data);
    return response.data;
  }

  async recordSignature(
    contractId: string,
    signerEmail: string,
    signatureImageUrl: string,
    ipAddress?: string
  ): Promise<ContractSignature> {
    const response = await apiClient.post<ContractSignature>(`/contracts/${contractId}/record-signature`, null, {
      params: { signerEmail, signatureImageUrl, ipAddress },
    });
    return response.data;
  }

  async declineSignature(contractId: string, signerEmail: string): Promise<ContractSignature> {
    const response = await apiClient.post<ContractSignature>(`/contracts/${contractId}/decline-signature`, null, {
      params: { signerEmail },
    });
    return response.data;
  }

  async getSignatures(contractId: string): Promise<ContractSignature[]> {
    const response = await apiClient.get<ContractSignature[]>(`/contracts/${contractId}/signatures`);
    return response.data;
  }

  async getPendingSignatures(contractId: string): Promise<ContractSignature[]> {
    const response = await apiClient.get<ContractSignature[]>(`/contracts/${contractId}/signatures/pending`);
    return response.data;
  }

  async getSignatureSummary(contractId: string): Promise<SignatureSummary> {
    const response = await apiClient.get<SignatureSummary>(`/contracts/${contractId}/signatures/summary`);
    return response.data;
  }

  // ===================== Templates =====================

  async createTemplate(data: CreateContractTemplateRequest): Promise<ContractTemplate> {
    const response = await apiClient.post<ContractTemplate>('/contracts/templates', data);
    return response.data;
  }

  async getTemplateById(id: string): Promise<ContractTemplate> {
    const response = await apiClient.get<ContractTemplate>(`/contracts/templates/${id}`);
    return response.data;
  }

  async getTemplates(page: number = 0, size: number = 20): Promise<Page<ContractTemplate>> {
    const response = await apiClient.get<Page<ContractTemplate>>('/contracts/templates', {
      params: { page, size },
    });
    return response.data;
  }

  async getActiveTemplates(page: number = 0, size: number = 20): Promise<Page<ContractTemplate>> {
    const response = await apiClient.get<Page<ContractTemplate>>('/contracts/templates/active', {
      params: { page, size },
    });
    return response.data;
  }

  async getTemplatesByType(type: ContractType): Promise<ContractTemplate[]> {
    const response = await apiClient.get<ContractTemplate[]>(`/contracts/templates/type/${type}`);
    return response.data;
  }

  async searchTemplates(search: string, page: number = 0, size: number = 20): Promise<Page<ContractTemplate>> {
    const response = await apiClient.get<Page<ContractTemplate>>('/contracts/templates/search', {
      params: { search, page, size },
    });
    return response.data;
  }

  async updateTemplate(id: string, data: CreateContractTemplateRequest): Promise<ContractTemplate> {
    const response = await apiClient.put<ContractTemplate>(`/contracts/templates/${id}`, data);
    return response.data;
  }

  async deleteTemplate(id: string): Promise<void> {
    await apiClient.delete(`/contracts/templates/${id}`);
  }

  async toggleTemplateActive(id: string): Promise<ContractTemplate> {
    const response = await apiClient.patch<ContractTemplate>(`/contracts/templates/${id}/toggle-active`);
    return response.data;
  }

  // ===================== Helpers =====================

  getTypeLabel(type: ContractType): string {
    const labels: Record<ContractType, string> = {
      EMPLOYMENT: 'Employment Contract',
      VENDOR: 'Vendor Contract',
      NDA: 'Non-Disclosure Agreement',
      SLA: 'Service Level Agreement',
      FREELANCER: 'Freelancer Agreement',
      OTHER: 'Other',
    };
    return labels[type] || type;
  }

  getStatusLabel(status: ContractStatus): string {
    const labels: Record<ContractStatus, string> = {
      DRAFT: 'Draft',
      PENDING_REVIEW: 'Pending Review',
      PENDING_SIGNATURES: 'Pending Signatures',
      ACTIVE: 'Active',
      EXPIRED: 'Expired',
      TERMINATED: 'Terminated',
      RENEWED: 'Renewed',
    };
    return labels[status] || status;
  }

  getStatusColor(status: ContractStatus): string {
    const colors: Record<ContractStatus, string> = {
      DRAFT: 'bg-[var(--bg-surface)] text-[var(--text-secondary)]',
      PENDING_REVIEW: 'bg-warning-100 text-warning-700 dark:bg-warning-900 dark:text-warning-300',
      PENDING_SIGNATURES: 'bg-warning-100 text-warning-700 dark:bg-warning-900 dark:text-warning-300',
      ACTIVE: 'bg-success-100 text-success-700 dark:bg-success-900 dark:text-success-300',
      EXPIRED: 'bg-danger-100 text-danger-700 dark:bg-danger-900 dark:text-danger-300',
      TERMINATED: 'bg-danger-100 text-danger-700 dark:bg-danger-900 dark:text-danger-300',
      RENEWED: 'bg-accent-100 text-accent-700 dark:bg-accent-900 dark:text-accent-300',
    };
    return colors[status] || colors.DRAFT;
  }

  formatCurrency(value: number | undefined, currency: string = 'USD'): string {
    if (value === undefined) return 'N/A';
    return new Intl.NumberFormat('en-IN', {
      style: 'currency',
      currency: currency,
    }).format(value);
  }

  formatDate(dateString: string): string {
    return new Date(dateString).toLocaleDateString('en-IN', {
      day: '2-digit',
      month: 'short',
      year: 'numeric',
    });
  }
}

export const contractService = new ContractService();
